patches/sitecustomize.py

```diff
@@ -1,20 +1,56 @@
+from google.auth import credentials
+from google.cloud import bigquery
+from google.cloud.bigquery._http import Connection
 import os
 
-# Monkey patches BigQuery client creation to use proxy.
+
+class KaggleKernelCredentials(credentials.Credentials):
+    def refresh(self, request):
+        print("Calling Kaggle.UserSecrets to refresh token.")
+        # Set self.token and self.expiry here.
+        raise NotImplementedError("Private BigQuery integration is not yet implemented.")
+
 kaggle_proxy_token = os.getenv("KAGGLE_DATA_PROXY_TOKEN")
-if kaggle_proxy_token:
-    from google.auth import credentials
-    from google.cloud import bigquery
-    from google.cloud.bigquery._http import Connection
+CONNECTION_BASE_URL = Connection.API_BASE_URL
+
 
-    Connection.API_BASE_URL = os.getenv("KAGGLE_DATA_PROXY_URL")
-    Connection._EXTRA_HEADERS["X-KAGGLE-PROXY-DATA"] = kaggle_proxy_token
+def monkeypatch_bq(bq_client, *args, **kwargs):
+    data_proxy_project = os.getenv("KAGGLE_DATA_PROXY_PROJECT")
+    bq_user_jwt = os.getenv("KAGGLE_BQ_USER_JWT")
+    specified_project = kwargs['project'] if 'project' in kwargs else None
+    # Use Data Proxy if user has specified to use the Kaggle project, or if
+    # there are no connected GCP accounts (to maintain backwards compatibility).
+    if bq_user_jwt is None and specified_project and specified_project.lower() != 'kaggle':
+        raise Exception("In order to query a private BigQuery project, please connect a GCP account. "
+                        "Otherwise specify 'kaggle' as the project to use Kaggle's public dataset BigQuery integration.")
+    use_data_proxy = (specified_project and specified_project.lower() == 'kaggle') or bq_user_jwt is None
```

Comment:
Continued from:

```
@vimota
vimota 3 days ago  Member
This kaggle "project" the user specifies is really more of just a way to tell us they want to use the Kaggle "free bigquery" Data Proxy even though they've connected a GCP account. If they haven't connected their GCP account (ie. bq_user_jwt is not present) then they don't need specify a 'kaggle' or any other project, we will default to using the data proxy.

More info here: #414 (comment)
```

since github won't let me comment on that.

I find this really confusing.  That comment also specifies itself that it is confusing.  Don't explain confusing things in a review comment.  Document them to start with, including their motivation.

In this case I would like to understand why we're so hard trying to hijack the standard bq client to fit all sorts of customizations.  At least the "I explicitly want to use the Kaggle bq" could be more clearly supported that way.

Reply:
I had originally preferred doing something like that as well - ie. from Kaggle import bq (kaggle proxy) vs normal import bq (using connected account), but opted for this because it avoided breaking users existing Kernels. Do you think it's better to take that approach even if it breaks existing users, I agree it'll be cleaner?

Reply:
@sebbov And to provide a bit more context, the reason we want to keep everything as close as possible to the code people would write outside of kernels is because part of the appeal of GCP integrations from Google's perspective is teaching users how to use GCP tools. The fewer changes they need to make between Kaggle and non-Kaggle code, the easier they'll be able to use GCP for their non-Kaggle work. Or at least that's the thinking. I understand there's always going to be some magic involved wherever we want to do that, but I think BQ is likely the worst offender since there's both the concept of a public BQ product and private BQ product rolled into the same bq library.

Reply:
@harrisse 👍 Does it sound right that this means:

- Let them use the vanilla BQ client for the non-proxy case. Provide an example on how to use the environment-specified credentials.
- For the proxy case, wrap this in a separate class (and note what you can't do in this case, e.g. the concurrent use w/ the vanilla client mentioned in a different comment) since there is no such thing as proxy support in the current BQ client at this time.